Set the timezone correctly.

toolstack opened the pull request:
This avoids daylight savings zone problems with future dates.

## What?
Fixes #46333, which shows the wrong time in Gutenberg if the post is in the future and crosses the daylight savings time boundries.

## Why?
See #46333, showing the incorrect publish time in the UI is bad.

## How?
Instead of using a fixed "WP" timezone, that is added to moment.tz's zones table, use the "proper" timezone as supplied by WP if possible.

## Testing Instructions
1. Make sure you are in a timezone that uses daylight savings time.
2. Create a post.
3. Set it to publish after the next daylight savings time change.
4. See the date reported in Gutenberg's summary widget  is off by an hour, but nowhere else (aka the datepicker is fine, and the post list is fine, and the post will be published on the date you entered in the datepicker).

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast
See #46333 for screenshots.

ajlende commented on packages/date/src/index.js:
```diff
@@ -57,7 +57,7 @@ import deprecated from '@wordpress/deprecated';
  */
 
-const WP_ZONE = 'WP';
+var WP_ZONE = 'WP';
```
Two code style things here:

1. We're writing modern JS in Gutenberg, so we use `let` or `const` instead of `var`. See the [Assignments and Globals > Declaring Variables with `const` and `let`](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#declaring-variables-with-const-and-let) section in the guide for details.
2. Variable names in all caps represents constants that don't change, and we use camelCase for everything else. See the [Naming Conventions > Constants](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#constants) section in the guide for details.

That being said, I don't think updating the `WP_ZONE` is the right way to solve this. I'll share why in the main review comment.

toolstack replied:
Fixed.

ajlende commented on packages/date/src/index.js:
```diff
@@ -145,6 +145,13 @@ export function setSettings( dateSettings ) {
 
+	// If we can match a real timezone string, set WP_ZONE to it so we handle
+	// daylight savings time properly for different regions that use it.
+	if (momentLib.tz.zoneExists(settings.timezone.string)) {
```
The errors listed suggest that you're using the wrong method for checking if a timezone exists. The moment docs for [checking zone existance](https://momentjscom.readthedocs.io/en/stable/moment-timezone/04-data-loading/04-checking-if-a-zone-exists/) say to use `zone()`, not `zoneExists()`.

Another code style thing is that we add spacing in lots of places like this. See the [Spacing](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#spacing) section for details.


```suggestion
	if ( momentLib.tz.zone( settings.timezone.string ) ) {
```

ajlende reviewed (changes requested):
Thanks for your contribution to Gutenberg!

---

I added a couple notes about our JS coding standards for future reference. The full guide is can be found in the developer docs under [JavaScript Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript).

We have a "Static Analysis (Linting, License, Type checks...)" check that runs on each pull request that enforces these coding standards. You can see the results at the bottom by clicking the "Details" link next to it. But you can also run them on your own computer with `npm run lint:js` so you don't have to push up code and wait for it to run here (which can take quite a while). There will be some warnings listed, but you only need to worry about any errors reported.

---

As far as the fix is concerned, I don't think updating the `WP_ZONE` is the right way to solve this.

It looks like `'WP'` is a custom timezone created in `setupWPTimezone()` with custom settings. I think it would be better if we could copy over the settings from the `settings.timezone.string` moment timezone in that function. That way, if someone wants to customize the timezone with `wp.date.setSettings()`, those customizations can override it.

toolstack commented:
> I added a couple notes about our JS coding standards for future reference.

Np, I wasn't too worried about the coding sytle at this point as I expected there to be some discussion on if this was even the right way to resolve the issue.

> As far as the fix is concerned, I don't think updating the `WP_ZONE` is the right way to solve this.
> 
> It looks like `'WP'` is a custom timezone created in `setupWPTimezone()` with custom settings. I think it would be better if we could copy over the settings from the `settings.timezone.string` moment timezone in that function. That way, if someone wants to customize the timezone with `wp.date.setSettings()`, those customizations can override it.

As I recall (it's been a while since I looked at the code) the WP custom zone is just a offset from UTC, which is the problem, as the rest of the timezone functions can no longer tell if daylight savings times should be applied or not.  Keeping the original timezone info as part of WP_ZONE allows the timezone code to determine if DST is required or not for that given zone.

For example, if you are in EST (EDT for DST), then you are at UTC -5, the WP timezone gets set to -5 and used in the editor.  But when a future date is used, the timezone functions can no longer tell that you're actually in EST instead of some other timezone that is -5 from UTC, and therefore there is no way to tell that EDT should be used instead.

I think timezone data should kinda be like santaization/escaping, validated early (aka make sure you have a real timeone) and then escaped late (aka don't apply the timezone offset to the date until just before display in case something has changed).  Using the WP timezone breaks the second part by effectively pre-calculating the timezone offset early during the validation and then never re-calculating it before display.

ajlende commented:
> Np, I wasn't too worried about the coding style at this point as I expected there to be some discussion on if this was even the right way to resolve the issue.

I forgot to mention earlier that we have a script that will just fix things so you don't have to worry about it: `npm run lint:js:fix`.

> As I recall (it's been a while since I looked at the code) the WP custom zone is just a offset from UTC

Can't the WP custom zone also include the information of when DST changes instead of just the offset?

I'm not all that familiar with `moment-timezone`, but I'd imagine custom zones should be able to include DST information just like the named zones. You'll have to read the moment docs to see how to do that.

toolstack committed:
Add additional information to the WP timezone.

Add additional information, like abbreviations and offsets from the moment timezone data to the custom WP timezone if available.

This avoids daylight savings zone problems with future dates.

toolstack commented:
> I forgot to mention earlier that we have a script that will just fix things so you don't have to worry about it: `npm run lint:js:fix`.

Which sounds good until I have to go down the rabbit hole of having npm v9... 😭😂

> Can't the WP custom zone also include the information of when DST changes instead of just the offset?
>
> I'm not all that familiar with `moment-timezone`, but I'd imagine custom zones should be able to include DST information just like the named zones. You'll have to read the moment docs to see how to do that.

The custom WP timezone structure does have the offsets array, which does look to be used for this purpose, but we'd have to effectively duplicate the work that moment does to manage this data.

I guess you could, instead of setting WP_ZONE, pull a copy of the data from the real timezone info from moment and then copy that into the WP custom timezone, but that seems like a lot of work for no real benefit.

I'm not sure why the custom WP timezone was used, I'm sure there was a valid reason to do so, I just think it adds a lot of that seems wasted.

toolstack commented:
> I guess you could, instead of setting WP_ZONE, pull a copy of the data from the real timezone info from moment and then copy that into the WP custom timezone, but that seems like a lot of work for no real benefit.

I've mocked up what that looks like:

```
function setupWPTimezone() {
  // Get the current timezone settings from the WP timezone string.
  let temp_tz = momentLib.tz.zone( settings.timezone.string );

  // Check to see if we have a valid TZ data, if so, use it fro the custom WP_ZONE timezone, otherwise just use the offset.
  if( temp_tz ) {

    // Create WP timezone based off dateSettings.
      momentLib.tz.add(momentLib.tz.pack({
      name: WP_ZONE,
      abbrs: temp_tz.abbrs,
      untils: temp_tz.untils,
      offsets: temp_tz.offsets
    }));

  } else {

    // Create WP timezone based off dateSettings.
    momentLib.tz.add(momentLib.tz.pack({
      name: WP_ZONE,
      abbrs: [WP_ZONE],
      untils: [null],
      offsets: [-settings.timezone.offset * 60 || 0]
    }));
  }
} // Date constants.
```

This does work, but the code is run twice, as setupWPTimezone() is run during the initial javascript load when `settings.timezone.string` hasn't been set yet, and a second time when `setSettings()` is called and the `settings.timezone.string` has been set.

ajlende commented:
> This does work, but the code is run twice, as `setupWPTimezone()` is run during the initial javascript load when `settings.timezone.string` hasn't been set yet, and a second time when `setSettings()` is called and the `settings.timezone.string` has been set.

Good. It's important to update the timezone info when `setSettings()` is called. Maybe someone in a plugin is calling it a third time and expecting it to update the settings.

> ```js
> // Create WP timezone based off dateSettings.
> momentLib.tz.add( momentLib.tz.pack( {
> 	name: WP_ZONE,
> 	abbrs: temp_tz.abbrs,
> 	untils: temp_tz.untils,
> 	offsets: temp_tz.offsets
> } ) );
>```

I checked the moment docs, and it looks like there's a better way to setup a new timezone with identical settings using the [`momentLib.tz.link()`](https://momentjscom.readthedocs.io/en/stable/moment-timezone/03-data-formats/03-link-format/) method.

```js
momentLib.tz.link( `${ settings.timezone.string }|${ WP_ZONE }` );
```

>```js
> function setupWPTimezone() {
> 	// Get the current timezone settings from the WP timezone string.
> 	const temp_tz = momentLib.tz.zone( settings.timezone.string );
> 
> 	// Check to see if we have a valid TZ data, if so, use it for the custom WP_ZONE timezone, otherwise just use the offset.
> 	if ( temp_tz ) {
> 		...
> 	} else {
> 		...
> 	}
> }

I also looked through a few places where this GMT offset vs timezone string is used in core<sup>[1](https://github.com/WordPress/wordpress-develop/blob/5f1a92c96da0b2e271c32c5d147a3847b8960a12/src/wp-includes/functions.php#L100-L136),[2](https://github.com/WordPress/wordpress-develop/blob/5f1a92c96da0b2e271c32c5d147a3847b8960a12/src/wp-includes/functions.php#L6442-L6464),[3](https://github.com/WordPress/wordpress-develop/blob/5f1a92c96da0b2e271c32c5d147a3847b8960a12/src/wp-admin/options-general.php#L233-L252)</sup> to confirm that the string should override the offset, and it does work that way everywhere I looked. So this conditional is correct. 👍 

---

@toolstack You're on the right track. Get these changes in, and we can get it merged. 🙂

toolstack commented:
> I checked the moment docs, and it looks like there's a better way to setup a new timezone with identical settings using the [`momentLib.tz.link()`](https://momentjscom.readthedocs.io/en/stable/moment-timezone/03-data-formats/03-link-format/) method.

Using ```link()``` fails as we've already created the WP timezone on the first pass through and it looks like you can't overwrite an existing timezone with a link, so I left it as it was as you can overwrite a timezone with another one with ```add()```.

ajlende commented on packages/date/src/index.js:
```diff
@@ -226,15 +226,31 @@ export function __experimentalGetSettings() {
-			offsets: [ -settings.timezone.offset * 60 || 0 ],
-		} )
-	);
+	// Get the current timezone settings from the WP timezone string.
+	const currentTimeZone = momentLib.tz.zone( settings.timezone.string );
+
+	// Check to see if we have a valid TZ data, if so, use it fro the custom WP_ZONE timezone, otherwise just use the offset.
+	if ( currentTimeZone ) {
+		// Create WP timezone based off dateSettings.
+		momentLib.tz.add(
+			momentLib.tz.pack( {
+				name: WP_ZONE,
+				abbrs: currentTimeZone.abbrs,
+				untils: currentTimeZone.untils,
+				offsets: currentTimeZone.offsets,
+			} )
+		);
+	} else {
+		// Create WP timezone based off dateSettings.
+		momentLib.tz.add(
+			momentLib.tz.pack( {
+				name: WP_ZONE,
+				abbrs: [ WP_ZONE ],
+				untils: [ null ],
+				offsets: [ -settings.timezone.offset * 60 || 0 ],
+			} )
+		);
+	}
```
Adding two minor suggestions:

1. Rename `currentTimeZone` to `currentTimezone` since everywhere else treats "timezone" as a single word instead of two.
2. Update comments to reflect discussion in this PR. This is more of a personal philosophy than a WP one, but I find comments that explain "why I wrote it this way" much more useful than ones that explain "what I wrote".

```suggestion
	const currentTimezone = momentLib.tz.zone( settings.timezone.string );

	// Prefer the timezone string over the offset like other places using both.
	if ( currentTimezone ) {
		// Use add() instead of link() because add() will override the
		// existing WP_ZONE if it is already set.
		momentLib.tz.add(
			momentLib.tz.pack( {
				name: WP_ZONE,
				abbrs: currentTimezone.abbrs,
				untils: currentTimezone.untils,
				offsets: currentTimezone.offsets,
			} )
		);
	} else {
		momentLib.tz.add(
			momentLib.tz.pack( {
				name: WP_ZONE,
				abbrs: [ WP_ZONE ],
				untils: [ null ],
				offsets: [ -settings.timezone.offset * 60 || 0 ],
			} )
		);
	}
```

toolstack replied:
Done.

toolstack committed:
Rename variable and add additional comments.

ajlende reviewed (approved):
LGTM. This is what I see. I set my timezone set to 'America/Chicago' which has DST changing on November 5.

| Before | After |
| ------ | ----- |
| <img width="310" alt="image" src="https://github.com/WordPress/gutenberg/assets/5129775/c323b28d-1020-4d7c-a05a-b07e2a7cd152"> | <img width="311" alt="image" src="https://github.com/WordPress/gutenberg/assets/5129775/8ba645eb-dff4-4e02-a994-e8b4170c5920"> |
